### PR TITLE
ADR: JavaScript Modules exports convention

### LINKED ISSUE #12: [Proposal] Avoid default export in JavaScript

### PR DESCRIPTION
**Pull request by yordis**
Please share your thoughts, concerns, feedbacks ... constructive criticism is welcome

closes #12

### CODE REVIEW

File: adrs/4937890790/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+```ts
+import [name here] from 'base-button'
+```
+
+What should be the right name for the import?
+
+* `Button`
+* `button`
+* `baseButton`
```
[the remark this answers is not in the document]
Comment (yordis) [reply adrs/4937890790/README.md:30]: @TMiller00 good observation, that is actually another reason why it is important to give named-exports. But let me remove that one.

File: adrs/4937890790/README.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+We don't have a consistent way to export the modules' functions and variables
+across the JavaScript ecosystem.
+
+Although such concern is not a big concern for more Senior developers, there is
+some cognitive load when it comes to default exports over named exports.
+
+What is one of the most challenging things in computer science: naming.
+
+Using default-exports over named-exports forces the user to think about the name
+of the import.
```
[the remark this answers is not in the document]
Comment (yordis) [reply adrs/4937890790/README.md:18]: > Although such concern is not a big concern for more Senior developers

😄 you are too ahead of most people, I am taking into consideration Bootcamp students, and self-taught people, and so on

File: adrs/6860374633/README.md [hunk 3]
```diff
@@ -116,7 +116,7 @@ The following example shows a CloudEvent Error serialized as JSON:
     "subject" : "/payment/amount",
     "datacontenttype" : "application/json",
     "data" : "{
-      \"errormessage\":\"You can not transfer an amount greater than the current balance\",
+      \"errormessage\":\"Can not transfer an amount greater than the current balance\",
```
Comment (yordis) [inline adrs/6860374633/README.md:119]: Ignore this, the linter is failing due to line character len

File: adrs/4937890790/README.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+* **Created:** 2021-02-17
+* **Tags:** js,ts
+
+## Context
+
+We don't have a consistent way to export a module's functions and variables
+across the JavaScript ecosystem.
+
+Although such concern is not a big concern for more Senior developers, there is
+some cognitive load when it comes to default exports over named exports.
```
Comment (manuphatak) [inline adrs/4937890790/README.md:13]:
```suggestion
We don't have a consistent way to export a module's functions and variables
across the JavaScript ecosystem.  In the modern era of ES6 exports, there are
generally 2 camps:

1. Use `export default` for your "main" export. 
  This option forces a second choice for auxiliary exports, where you can either 
  use named exports or `Object.assign()` the default.
2. Use named exports for everything.

Teams that don't choose a camp are inflicting an unnecessary cognitive load 
on themselves.
```

Wanted to take a stab at laying the options.

File: adrs/4937890790/README.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+What should be the right name for the import?
+
+* `Button`
+* `baseButton`
+* `BaseButton`
+* `MyButton` or any other name you would like it to be
+
+For a simple import, you have already at least three different variations to
+think about it, or combinations of any name for that matter.
+
```
Comment (manuphatak) [inline adrs/4937890790/README.md:35]: I know this comes up later, but I think it's horribly BM to rename imports -- even  in the `export default` world.  😂 


```ts
import Button from 'base-button'
```




In some cases, you have to avoid name conflict.

```ts
import TheirBaseButton from 'base-button'
import OurBaseButton from '../base-button'
```


Besides that, why would anybody do that?

File: adrs/4937890790/README.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+What should be the right name for the import?
+
+* `Button`
+* `baseButton`
+* `BaseButton`
+* `MyButton` or any other name you would like it to be
+
+For a simple import, you have already at least three different variations to
+think about it, or combinations of any name for that matter.
```
Comment (manuphatak) [inline adrs/4937890790/README.md:34]:
```suggestion
For a simple default import, you already have at least three different variations to
think about and combinations of any name for that matter.
```

File: adrs/4937890790/README.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+`PascalCase.` That is a limitation of React itself. You can not be "100% sure"
+unless you inspect the file and understand the context of such default-export.
+
+It is not a simple task; therefore, people rely on linters and code reviewers to
+improve the source code's quality.
+
+If you already thought about a name and knew the context, and gave it a name
+inside your module, don't force users to think about a new name every time
+they must import the component, most of the case the named you came up with is
+good enough; so you can take advantage of named-exports.
```
Comment (manuphatak) [inline adrs/4937890790/README.md:50]:
```suggestion
Naming is hard.  When you are authoring code, do not make the consumers of your
code -- the importers of your exports --  figure out what names to use.  As an author, 
you have the advantage of having thought about its intent and purpose; you are
aware of its context and limitations and most likely gave it a name inside your module.
That name is good enough, and by using named exports, you  are doing your consumers a favor, 
by freeing them from having to rediscover that name. 
```

File: adrs/4937890790/README.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+`Button` context, because you still need to know the import, you are somehow
+optimistic that it may be.
+There is an explicit remapping happening, helping the reader to understand that
+in this file, for whatever reason, `Button` must become `Btn`.
+
+Also, having explicit names allows IDEs to figure out what you are trying to
+accomplish and give you much better autocomplete, refactoring, and automation,
+since those named-exports are static and don't require runtime to figure out
+the names, they are able to index the source code much easier and algorithms
+could take advantage of the explicitness and static nature of such habit.
```
Comment (manuphatak) [inline adrs/4937890790/README.md:144]:
```suggestion
Also, having explicit names allows IDEs to figure out what you are trying to
accomplish and give you much better autocomplete refactoring and automation.
Since those named-exports are static and don't require a runtime to figure out
the names, they can index the source code and algorithms
can take advantage of the explicitness and static nature of the names.
```

File: adrs/4937890790/README.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+## Resolution
+
+* You **must** use named-exports over default-exports unless some limitation
+  is imposed
+* You **must** provide named-export and default-exports for React lazy
+  loadable components.
```
Comment (manuphatak) [inline adrs/4937890790/README.md:151]: What do you think about adding a resolution, something like:

You should not rename code unless you're solving a scoping conflict.

**Review by manuphatak — APPROVED** [review]
🙌   Awesome work! I love these ADRs.


I rewrote a couple of sections. Definitely don't feel obligated to use them.  In a couple of cases, these were my reflections on the section. 👍

**Comment by manuphatak** [comment]
In case you want the ADR to address the strongest arguments for "default exports", let me take a stab at laying out what that is.

I was a big believer in "default exports".  Even though I abandoned them, I still believe "default exports" is an expression of the "single responsibility principle". 

Under this line of thinking, the resolution would've been:
-  "a module **MUST** represent a single, exportable idea."

In other words, if you feel the need to export multiple things, you have NOT done the work of discovering what that modules single responsibility is.

**Comment by yordis** [comment]
> In other words, if you feel the need to export multiple things, you have NOT done the work of discovering what that modules single responsibility is.

You had the answer almost, the module itself is the single responsibility principle 🙂

Exporting multiple things, although it may feel that is doing multiple thing, it is not. Especially if your programming language itself is not pure object oriented like Java or C#.

Now, it your module is all over the place, then I resonate with your concern but that is another conversation outside of this ADR scope.

Your concerns are there regardless of the strategy of how you export things, for example, I can export a single object (as it is done in CJS) wiht many things in it and have the same concern. Or even export a class (an object for that matter) and have a lot of static functions in it just because.

So I totally agree with your concern but your are feeling weird about such thing for the wrong reasons I think.

does that makes sense?

File: adrs/4937890790/README.md (hunk 9 above)
Comment (yordis) [reply adrs/4937890790/README.md:151]: I am not following, what do you mean by scoping conflict? and where?

**Comment by yordis** [comment]
Moved the ADR to reviewing

**Comment by tleef** [comment]
👍
